crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -578,7 +579,54 @@ impl RoomSendQueue {
                 continue;
             };
 
-            match Self::handle_request(&room, queued_request, cancel_upload_rx).await {
+            let is_thumbnail = match queued_request.kind {
+                QueuedRequestKind::MediaUpload { is_thumbnail, .. } => is_thumbnail,
+                _ => false,
+            };
+
+            let index = {
+                cfg_if::cfg_if! {
+                    if #[cfg(feature = "unstable-msc4274")] {
+                        match &queued_request.kind {
+                            QueuedRequestKind::MediaUpload { accumulated, .. } => accumulated.len() as u64,
+                            _ => 0,
+                        };
+                    } else {
+                        0 // Before MSC4274 there were only a single file (and thumbnail) could be sent per event.
+                    }
+                }
+            };
+
+            let progress_watcher = {
+                if let Some((progress, related_to)) = match &queued_request.kind {
+                    QueuedRequestKind::MediaUpload { related_to, .. } => {
+                        let progress: SharedObservable<TransmissionProgress> = Default::default();
+                        Some((progress, related_to.clone()))
+                    }
+                    _ => None,
+                } {
+                    let mut subscriber = progress.subscribe();
+                    let our_updates = updates.clone();
+                    spawn(async move {
```

Comment:
Who controls this task? If the upload is cancelled, this task should be cancelled too right? Or maybe the `subscriber.next().await` may return a `None` at some point? If that's the case, please test that behaviour specifically, and comment that out, otherwise handle this 🙂.

Reply:
Yes, indeed `next()` will yield `None` once the observable (`progress`) and its clones are dropped. This happens once the final `RoomSendQueueUpdate::MediaUpload` is emited further down after the request has completed.

I've added a comment about this but could also add an explicit `abort` if that would be better.